File: src/line_striping.cpp
#include <ros/ros.h>
#include <ros/console.h>
#include <cv_bridge/cv_bridge.h>

#include <nav_msgs/Odometry.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/image_encodings.h>

#include <opencv2/imgproc.hpp>
#include <opencv2/ximgproc.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

ros::Publisher pointcloud_pub;
nav_msgs::Odometry odom;

void odomCallback(const nav_msgs::Odometry::ConstPtr &odom_msg) {
  odom.header = odom_msg->header;
  odom.child_frame_id = odom_msg->child_frame_id;
  odom.pose = odom_msg->pose;
  odom.twist = odom_msg->twist;
}

void leftCallback(const sensor_msgs::Image::ConstPtr &left_msg) {
  cv_bridge::CvImagePtr left_ptr;
  try {
    left_ptr = cv_bridge::toCvCopy(left_msg, sensor_msgs::image_encodings::BGR8);
  } catch (cv_bridge::Exception& e) {
    ROS_ERROR("cv_bridge exception: %s", e.what());
    return;
  }

  cv::Mat hsv, thresholded, eroded;
  cv::Mat dilated1, dilated2, dilated3;
  cv::Mat thinned1, thinned2;
  cv::Mat output;

  cv::Mat4i lines;

  cv::cvtColor(left_ptr->image, hsv, cv::COLOR_BGR2HSV);
  cv::inRange(hsv, cv::Scalar(70, 30, 180), cv::Scalar(100, 255, 255), thresholded);
  cv::ximgproc::thinning(thresholded, thinned1, cv::ximgproc::THINNING_ZHANGSUEN);
  cv::dilate(thinned1, dilated1, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5)));
  cv::dilate(dilated1, dilated2, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5)));
  cv::dilate(dilated2, dilated3, cv::getStructuringElement(cv::MORPH_RECT, cv::Size(5, 5)));
  cv::ximgproc::thinning(dilated3, thinned2, cv::ximgproc::THINNING_ZHANGSUEN);
  
  cv::Ptr<cv::LineSegmentDetector> lsd = cv::createLineSegmentDetector();
  lsd->detect(thinned2, lines);
  lsd->drawSegments(output, lines);

  cv::imshow("Image1", output);
  cv::waitKey(3);
}

int main(int argc, char** argv) {
  ros::init(argc, argv, "line_striping_node");
  ros::NodeHandle nh;

  ros::Subscriber odom_sub = nh.subscribe("/odom", 1, odomCallback);
  ros::Subscriber left_sub = nh.subscribe("/mipi/cam0", 1, leftCallback);
  //ros::Subscriber right_sub = nh.subscribe("/mipi/cam1", 1, rightCallback);

  //output_cloud->header.frame_id = "base_link";
  //pointcloud_pub = nh.advertise<PointCloud<PointXYZ>>("points", 1);

  ros::spin();
  return 0;
}

File: src/line_striping.py
#!/usr/bin/env python

import cv2
import math
import numpy as np

import rospy
from cv_bridge import CvBridge
from sensor_msgs import point_cloud2

from std_msgs.msg import Header
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Image
from sensor_msgs.msg import PointCloud2

SENSOR_HEIGHT  = 0.00356
SENSOR_WIDTH   = 0.00626
FOCAL_LENGTH   = 0.005
BASELINE       = 3 * 0.0254
CAMERA_ANGLE   = -35

LINE_POINTS    = ((0, 639), (719, 639))
WINDOW_SIZE    = 10
PLANE_EQUATION = (0.99978033, -0.00388811, 0.02059571, -0.03468802)

bridge    = CvBridge()
odom_msg  = Odometry()
left_msg  = Image()
right_msg = Image()

pointcloud_pub = rospy.Publisher('/points', PointCloud2, queue_size=1)

def odomCallback(msg):
  odom_msg = msg

def leftCallback(msg):
  left_msg = msg

def rightCallback(msg):
  right_msg = msg

  img_l = bridge.imgmsg_to_cv2(left_msg, desired_encodings='passthrough')
  img_r = bridge.imgmsg_to_cv2(right_msg, desired_encodings='passthrough')
  rows, cols, _ = img_l.shape
  points = []

  ind_l = LINE_POINTS[0][1]
  ind_r = LINE_POINTS[0][1]
  for i in range(rows):
    max_l = -1
    max_r = -1
    for j in range(WINDOW_SIZE):
      sl = ind_l - (WINDOW_SIZE / 2) + j
      sr = ind_r - (WINDOW_SIZE / 2) + j
      if img_l[i, sl][1] > max_l:
        max_l = img_l[i, sl][1]
        ind_l = sl
      if img_r[i, sr][1] > max_r:
        max_r = img_r[i, sr][1]
        ind_r = sr

    if max_l == -1:
      ratio = (float(i + 1) / rows)
      ind_l = int(((1.0 - t) * LINE_POINTS[0][1]) + t * LINE_POINTS[1][1])
      for j in range(WINDOW_SIZE):
        sl = ind_l - (WINDOW_SIZE / 2) + j
        if img_l[i, sl][1] > max_l:
          max_l = img_l[i, sl][1]
          ind_l = sl

    if max_r == -1:
      ratio = (float(i + 1) / rows)
      ind_r = int(((1.0 - t) * LINE_POINTS[0][1]) + t * LINE_POINTS[1][1])
      for j in range(WINDOW_SIZE):
        sr = ind_r - (WINDOW_SIZE / 2) + j
        if img_r[i, sr][1] > max_r:
          max_r = img_r[i, sr][1]
          ind_r = sr

    x_l = (float(ind_l + 1) / cols) * SENSOR_W
    x_r = (float(ind_r + 1) / cols) * SENSOR_W
    y_l = ((float(i - (rows / 2)) - 0.5) / rows) * SENSOR_H
    y_r = ((float(i - (rows / 2)) - 0.5) / rows) * SENSOR_H

    z = (-PLANE_EQUATION[3] * FOCAL_LENGTH) / ((PLANE_EQUATION[0] * x_r) \
        + (PLANE_EQUATION[1] * y_r) + (PLANE_EQUATiON[2] * FOCAL_LENGTH))
    x = (x_r * z) / FOCAL_LENGTH
    y = (y_r * z) / FOCAL_LENGTH

    px = (points[i][1] * math.sin(ANGLE * math.pi / 180)) \
       + (points[i][2] * math.cos(ANGLE * math.pi / 180))
    pz = (points[i][1] * math.cos(ANGLE * math.pi / 180)) \
       - (points[i][2] * math.sin(ANGLE * math.pi / 180))
    py = x

    points.append([px, py, pz])
  
  header = Header()
  header.frame_id = "map"
  pc2 = point_cloud2.create_cloud_xyz32(header, points)
  pointcloud_pub.publish(pc2)

def run():
  rospy.init_node('line_striping_node', anonymous=True)
  rospy.Subscriber('/odom', Odometry, odomCallback)
  rospy.Subscriber('/mipi/cam0', Image, leftCallback)
  rospy.Subscriber('/mipi/cam1', Image, rightCallback)
  rospy.spin()

if __name__=='__main__':
  try:
    run()
  except rospy.ROSInterruptException:
    pass
